domehunter/dome_control.py

```diff
@@ -0,0 +1,674 @@
+
+    def _simulate_ticks(self, num_ticks):
+        """
+        Method to simulate encoder ticks while in testing mode.
+        """
+        tick_count = 0
+        # repeat this loop of driving the mock pins low then high to simulate
+        # an encoder tick. Continue until desired number of ticks is reached.
+        while tick_count < num_ticks:
+            self.encoder_pin.drive_low()
+            # test_mode_delay_duration is set so that it will always exceed
+            # the set bounce_time
+            # of the pins
+            time.sleep(self.test_mode_delay_duration)
+            self.encoder_pin.drive_high()
+            time.sleep(self.test_mode_delay_duration)
+            tick_count += 1
+
+    def _turn_led_on(self, leds=[]):  # pragma: no cover
+        """
+        Method of turning a set of debugging LEDs on
+
+        Parameters
+        ----------
+        leds : list
+            List of LED name string to indicate which LEDs to turn on.
+
+        """
+        # pass a list of strings of the leds to turn on
+        if not(self.debug_lights):
+            return None
+        if leds == []:
+            # if leds is an empty list do nothing
+            pass
+        # this function needs a bunch of checks at some point
+        # like length of the binary number, whether things have the right
+        # type at the end (binary int vs string vs list) etc etc
+        #
+        # take the current led_status and convert to a string in binary
+        # format (18bit)
+        new_state = format(self.led_status, '#020b')
+        # from that string create a list of characters
+        # use the keys in the leds list and the led_lights_ind
+        new_state = list(new_state)
+        for led in leds:
+            ind = self.led_lights_ind[led]
+            new_state[ind] = '1'
+        # convert the updated list to a string and then to a binary int
+        new_state = ''.join(new_state)
+        self.led_status = int(new_state, 2)
+        # pass the new binary int to LED controller
+        sn3218.enable_leds(self.led_status)
+
+    def _turn_led_off(self, leds=[]):  # pragma: no cover
+        """
+        Method of turning a set of debugging LEDs off.
+
+        Parameters
+        ----------
+        leds : list
+            List of LED name string to indicate which LEDs to turn off.
+
+        """
+        # pass a list of strings of the leds to turn off
+        if not(self.debug_lights):
+            return None
+        if leds == []:
+            # if leds is an empty list do nothing
+            pass
+        # take the current led_status and convert to a string in binary
+        # format (18bit)
+        new_state = format(self.led_status, '#020b')
+        # from that string create a list of characters
+        # use the keys in the leds list and the led_lights_ind
+        new_state = list(new_state)
+        for led in leds:
+            ind = self.led_lights_ind[led]
+            new_state[ind] = '0'
```

Comment:
Lots of repetition between the on and off methods for the leds and as far as I can see this line is the only difference. Can create one method that does all this work (e.g. `_change_led_state(0)` for off and pass `1` for on) and then the on and off just call that one method.